backend-api/src/functions/adapters/dynamoDbAdapter.ts

```diff
@@ -0,0 +1,195 @@
+    requestHandler: new NodeHttpHandler({
+      connectionTimeout: 29000,
+      requestTimeout: 29000,
+    }),
+  });
+
+  constructor(tableName: string) {
+    this.tableName = tableName;
+  }
+
+  async readSessionId(
+    subjectIdentifier: string,
+  ): Promise<Result<string | null>> {
+    const currentTimeInSeconds = Math.floor(Date.now() / 1000);
+    const input: QueryCommandInput = {
+      TableName: this.tableName,
+      IndexName: "subjectIdentifier-timeToLive-index",
+      KeyConditionExpression:
+        "subjectIdentifier = :subjectIdentifier and :currentTimeInSeconds < timeToLive",
+      FilterExpression: "sessionState = :sessionState",
+      ExpressionAttributeValues: {
+        ":subjectIdentifier": { S: subjectIdentifier },
+        ":sessionState": { S: "ASYNC_AUTH_SESSION_CREATED" },
+        ":currentTimeInSeconds": {
+          N: currentTimeInSeconds.toString(),
+        },
+      },
+      ProjectionExpression: "sessionId",
+      Limit: 1,
+      ScanIndexForward: false,
+    };
+
+    const queryResponse = await this.query(input);
+    if (queryResponse.isError || queryResponse.value === null) {
+      return queryResponse;
+    }
+
+    const record: DynamoDbRecord = queryResponse.value;
+    const sessionId = record.sessionId?.S;
+    if (!sessionId) {
+      return errorResult({
+        errorMessage: "Session is malformed",
+        errorCategory: "SERVER_ERROR",
+      });
+    }
+
+    return successResult(sessionId);
+  }
+
+  async readSessionDetails(
+    subjectIdentifier: string,
+  ): Promise<Result<SessionDetails | null>> {
+    const currentTimeInSeconds = Math.floor(Date.now() / 1000);
+    const input: QueryCommandInput = {
+      TableName: this.tableName,
+      IndexName: "subjectIdentifier-timeToLive-index",
+      KeyConditionExpression:
+        "subjectIdentifier = :subjectIdentifier and :currentTimeInSeconds < timeToLive",
+      FilterExpression: "sessionState = :sessionState",
+      ExpressionAttributeValues: {
+        ":subjectIdentifier": { S: subjectIdentifier },
+        ":sessionState": { S: "ASYNC_AUTH_SESSION_CREATED" },
+        ":currentTimeInSeconds": {
+          N: currentTimeInSeconds.toString(),
+        },
+      },
+      ExpressionAttributeNames: { "#state": "state" },
+      ProjectionExpression: "sessionId, redirectUri, #state",
+      Limit: 1,
+      ScanIndexForward: false,
+    };
+
+    const queryResponse = await this.query(input);
+    if (queryResponse.isError || queryResponse.value === null) {
+      return queryResponse;
+    }
+
+    const record: DynamoDbRecord = queryResponse.value;
+    const sessionId = record.sessionId?.S;
+    const state = record.state?.S;
+    if (!sessionId || !state) {
+      return errorResult({
+        errorMessage: "Session is malformed",
+        errorCategory: "SERVER_ERROR",
+      });
+    }
+    const redirectUri: string | undefined = record.redirectUri?.S;
+    const sessionDetails: SessionDetails = {
```

Comment:
minor one - I think this could just be Session?

Reply:
Changed to Session since we have either sessionId or session details, so the latter can just be session :D